### ISSUE TITLE
Table column management demo

### ISSUE THREAD

**Issue by mcarrano** [patternfly/patternfly-next#2626]
Introduce a demo that illustrates Table Column Management. This will be implemented in a modal that can be opened from the data toolbar when supported for tables. The dialog supports the ability to toggle column visibility and/or reorder columns by dragging and dropping them. Design here: https://docs.google.com/document/d/1QRCmAHcSrJ4uPlQsUiRLcezz9wo2zdIdAML1wBLTXYI/edit?usp=sharing

cc @mcoker @rachael-phillips @LHinson

**Comment by mcoker** [comment]
@mcarrano before proceeding with a demo on the website, here is a demo using the data-list component  as we discussed https://codepen.io/mcoker/pen/RwNOQNR

Some notes:
* I introduced negative margins on the modal body element wrapping the data-list since we currently we do not have a way to allow content in the modal body to extend to the left/right edges of the modal. We would need to find a way to introduce that if we proceed with the borders extending to the edges of the modal.
* We don't have a way in the modal to create space between the description and the data-list, so I wrapped the description and data-list in separate modal body elements, then configured the data-list to create space between 2 adjacent modal body elements. I think that's a fair enhancement to the modal component outside of this demo, too. We support that same behavior in other components, such as the card, when using multiple body elements.
* I modified the top/bottom padding of the data-list rows to be 16px. This could be achieved by adding a compact variation of the data-list component.

**Comment by mcarrano** [comment]
@mcoker Thanks for providing this demo. @maryshak1996 can you also take a look? Some thoughts on the above...

* While the design did have the list extending to the end of the modal, per design, I'm wondering if that's necessary. I think part of the problem may be that without some type of containing border, it might appear to be floating. Thoughts @maryshak1996 ?
* It should definitely be possible to have multiple elements in a modal like this. Didn't realize that was a problem. @mcoker do you think this calls for an update to the modal or is the technique you used adequate. The only think I don't understand here is why the description above the list gets a scroll bar. I would expect that to be fixed and the contents of the list to scroll if it overflows the height of the modal.
* Yeah, I was also thinking that we may need a compact variation of data-list to do this. Should that be opened as a separate issue?

**Comment by maryshak1996** [comment]
@mcoker this is looking great! I definitely don't have strong feelings about if the rows span full width or not (I'd be fine wit them just going to the edge of the inner padding of the modal as well). @mcarrano I definitely see what you mean about the checkboxes seeming misaligned -- so if making the rows NOT full width fixes that issue, then that works for me 
![Screen Shot 2020-01-28 at 11 28 58 AM](https://user-images.githubusercontent.com/19824574/73284034-f863b900-41c1-11ea-8090-33b567f25361.png)

**Comment by mcoker** [comment]
@mcarrano 

> do you think this calls for an update to the modal or is the technique you used adequate. 

This would be an update to the modal to support multiple body elements being stacked to define separate sections in the modal body.

> The only think I don't understand here is why the description above the list gets a scroll bar. I would expect that to be fixed and the contents of the list to scroll if it overflows the height of the modal.

This would be another problem we would need to solve, and it has to do with the way that the modal body element scrolls to contain its overflow. Traditionally we have only supported a single body element in the modal, and if it has overflow, it scrolls if the modal height is too tall for the viewport. Since we would be introducing a feature for multiple modal body elements, we would need to look at handling that overflow differently so that each body element doesn't scroll independently.

One option could just be to allow for a variation on the modal body element to disable the scroll/overflow behavior, so that in this case, you would enable that variation on the paragraph body in this demo. Then the paragraph body element wouldn't scroll, but the data-list body element would.

Another option could be to add an additional element to the modal called "description" (or something like that), that is a small blurb that would appear below the modal title and _wouldn't scroll_ if the contents of the description overflow.

> Yeah, I was also thinking that we may need a compact variation of data-list to do this. Should that be opened as a separate issue?

Yes.

@maryshak1996 

> the checkboxes seeming misaligned

Ah nice catch! That is due to how the datalist's (and table and other components) outer padding changes from 24px to 16px responsively between desktop and mobile. The modal component doesn't do that. What do you think about updating the modal so that it's outer padding adjusts to 16px on mobile screens? That should resolve the alignment issue you're seeing.

**Comment by maryshak1996** [comment]
@mcoker ya that sounds like a good solution to the alignment issue!

**Comment by mcarrano** [comment]
This article posted by @jessiehuff here gives some great advice on how to implement the drag and drop in the list to reorder items: https://github.com/patternfly/patternfly-design-kit/issues/154#issuecomment-579829385  See the section on resorting a list of items.

**Comment by maryshak1996** [comment]
@mcoker, yes what you have in that screenshot is what I would envision if we decided to go a route of not having full-width. 
And I think that I like the option of having this description section, to almost pin the description to the top with the header. In the codepen, the description is scrollable too and this seems strange to me, and not very intuitive:
<img width="595" alt="Screen Shot 2020-01-30 at 7 58 22 AM" src="https://user-images.githubusercontent.com/19824574/73451930-c031b700-4336-11ea-9faa-88580322cee1.png">

**Comment by maryshak1996** [comment]
Here is my final documentation around this demo: https://docs.google.com/document/d/1QRCmAHcSrJ4uPlQsUiRLcezz9wo2zdIdAML1wBLTXYI/edit?usp=sharing

**Comment by mcoker** [comment]
@mcarrano just to confirm, we have support for the toggle visibility demo, but not draggable rows in the data-list, so we can't build the column reordering demo yet. We will need an issue to add the draggable rows feature to the data list once the design is complete for that, and after we complete that, we can add the column reordering demo.

This demo is slated for our upcoming milestone - should we create a new issue to track the toggle visibility demo for this sprint? Or should we use this issue, and create a new one to track the draggable rows/column reordering demo?

**Comment by mcarrano** [comment]
@mcoker Yes, that is also what I expected. I've created https://github.com/patternfly/patternfly-next/issues/2728 for adding draggable rows to the data list with the understanding that this will be needed before we can support column reordering in the table. Once that is done the demo can be updated.

@LHinson does this sounds right to you? We probably need to add to the roadmap. Design should be complete based on @maryshak1996 's doc.
